x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/LifecyclePolicy.java

```diff
@@ -0,0 +1,150 @@
+            Map<String, Phase> phaseMap = phases.stream().collect(Collectors.toMap(Phase::getName, Function.identity()));
+            return new LifecyclePolicy(name, phaseMap);
+        });
+    private static Map<String, Set<String>> ALLOWED_ACTIONS = new HashMap<>();
+
+    static {
+        PARSER.declareNamedObjects(ConstructingObjectParser.constructorArg(), (p, c, n) -> Phase.parse(p, n), v -> {
+            throw new IllegalArgumentException("ordered " + PHASES_FIELD.getPreferredName() + " are not supported");
+        }, PHASES_FIELD);
+
+        ALLOWED_ACTIONS.put("hot", Sets.newHashSet(RolloverAction.NAME));
+        ALLOWED_ACTIONS.put("warm", Sets.newHashSet(AllocateAction.NAME, ForceMergeAction.NAME, ReadOnlyAction.NAME, ShrinkAction.NAME));
+        ALLOWED_ACTIONS.put("cold", Sets.newHashSet(AllocateAction.NAME));
+        ALLOWED_ACTIONS.put("delete", Sets.newHashSet(DeleteAction.NAME));
+    }
+
+    private final String name;
+    private final Map<String, Phase> phases;
+
+    /**
+     * @param name
+     *            the name of this {@link LifecyclePolicy}
+     * @param phases
+     *            a {@link Map} of {@link Phase}s which make up this
+     *            {@link LifecyclePolicy}.
+     */
+    public LifecyclePolicy(String name, Map<String, Phase> phases) {
+        this.name = name;
+        this.phases = phases;
+        validate(phases.values());
+    }
+
+    public static LifecyclePolicy parse(XContentParser parser, String name) {
+        return PARSER.apply(parser, name);
+    }
+
+    /**
+     * @return the name of this {@link LifecyclePolicy}
+     */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * @return the {@link Phase}s for this {@link LifecyclePolicy} in the order
+     *         in which they will be executed.
+     */
+    public Map<String, Phase> getPhases() {
+        return phases;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.startObject(PHASES_FIELD.getPreferredName());
+        for (Phase phase : phases.values()) {
+            builder.field(phase.getName(), phase);
+        }
+        builder.endObject();
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(name, phases);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null) {
+            return false;
+        }
+        if (obj.getClass() != getClass()) {
+            return false;
+        }
+        LifecyclePolicy other = (LifecyclePolicy) obj;
+        return Objects.equals(name, other.name) &&
+            Objects.equals(phases, other.phases);
+    }
+
+    @Override
+    public String toString() {
+        return Strings.toString(this, true, true);
+    }
+
+    static void validate(Collection<Phase> phases) {
```

Comment:
I think we can make this private? It shouldn't be too hard to test this method via the constructor?

Reply:
for sure! I'll just get rid of this method entirely